streaming/base/dataset.py

```diff
@@ -395,13 +395,17 @@ def __init__(self,
         self.shards_per_stream = np.zeros(self.num_streams, np.int64)
         self.sample_offset_per_stream = np.zeros(self.num_streams, np.int64)
         self.samples_per_stream = np.zeros(self.num_streams, np.int64)
+        self.max_shard_size_across_all_streams = 0
         for stream_id, stream in enumerate(self.streams):
             stream_shards = stream.get_shards(world)
             num_stream_samples = sum(map(len, stream_shards))
             if not num_stream_samples:
                 index_filename = os.path.join(stream.local, stream.split, get_index_basename())
                 raise RuntimeError(f'Stream contains no samples: {index_filename}.')
             stream_per_shard += [stream_id] * len(stream_shards)
+            # Includes the size of raw (decompressed) and zip (compressed) shard files.
+            self.max_shard_size_across_all_streams = max(self.max_shard_size_across_all_streams,
+                                                         stream_shards[0].get_max_size())
```

Comment:
there's a weird edge case here that honestly might not be worth addressing -- you might merge the index.json files from two different streams where stream A has shard file size 5x smaller than stream B, and the first shard in the combined index.json file is from stream A and it becomes the max shard in this loop. Then, we will still hang when trying to download any shard from stream B. This is pretty obscure though... see later comment for a fix

Reply:
No, that's a valid use-case. The only reason I did not wanted to iterate all the shards is because of the time it takes to perform the whole operation. But your use-case is a valid one.